8253079: DeterministicDump.java fails due to garbage in structure padding

iklam opened the pull request:
(EDITED) In product builds, when `PackageEntry` and `ModuleEntry` objects are allocated, the memory is not zeroed. As a result, the structure padding slots (such as the 32-bits after `BasicHashtableEntry::_hash`) may contain garbage values that are different on every run of `java -Xshare:dump`. As a result, `java -Xshare:dump` cannot reproduce deterministic result.

The fix is to clear the memory for the newly allocated `HashtableEntry` objects when `DumpSharedSpaces == true`.
<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253079](https://bugs.openjdk.java.net/browse/JDK-8253079): DeterministicDump.java fails due to garbage in structure padding


### Reviewers
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**) ⚠️ Review applies to 225c92728584527321e1cf199a31577ea5e31a6a
 * [Jie Fu](https://openjdk.java.net/census#jiefu) (@DamonFool - Committer)
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/267/head:pull/267`
`$ git checkout pull/267`

yminqi commented:
The implementation is quite complex, every derivatives from BasicHashTableEntry need to have a copy_from function to avoid such problem. Maybe a brutal one to avoid such padding issue for all cases is fill the object allocated in AllocateHeap with \0?

iklam commented:
> The implementation is quite complex, every derivatives from BasicHashTableEntry need to have a copy_from function to avoid such problem. Maybe a brutal one to avoid such padding issue for all cases is fill the object allocated in AllocateHeap with \0?

You're right. My original analysis was wrong: set_hash() didn't write garbage into the padding. Instead, the garbage was there because AllocaHeap didn't initialize the new buffer in product builds.

I reverted the original fix. Instead, I added code to call memset() when allocating a new hashtable entry (but only when DumpSharedSpaces is true).

shipilev commented on src/hotspot/share/classfile/moduleEntry.cpp:
```diff
@@ -454,6 +454,7 @@ void ModuleEntry::init_as_archived_entry() {
   if (_location != NULL) {
     _location = ArchiveBuilder::get_relocated_symbol(_location);
   }
+  JFR_ONLY(memset(trace_id_addr(), 0, sizeof(traceid)));
```
`memset` looks dodgy here. Maybe `JFR_ONLY(set_trace_id(0))`?

iklam replied:
I removed these from my previous commit, because the trace_id seemed to be predictable and wouldn't cause the archive content to change. Anyway, I've added the code you suggested in commit [5706a82](https://github.com/openjdk/jdk/pull/267/commits/5706a821a4b96c1c44c44ccf6d78c0659a6f2976). That way, if trace_ids become unpredictable due to future JFR implementation changes, the CDS code won't be affected.

shipilev commented on src/hotspot/share/classfile/packageEntry.cpp:
```diff
@@ -238,6 +238,7 @@ void PackageEntry::init_as_archived_entry() {
   _module = ModuleEntry::get_archived_entry(_module);
   _qualified_exports = (GrowableArray<ModuleEntry*>*)archived_qualified_exports;
   _defined_by_cds_in_class_path = 0;
+  JFR_ONLY(memset(trace_id_addr(), 0, sizeof(traceid)));
```
Ditto, `JFR_ONLY(set_trace_id(0))`?

DamonFool commented:
It passed with this fix.
Looks good to me.
Thanks.

DamonFool commented:
/reviewer jiefu

yminqi reviewed (approved):
looks good!

iklam commented:
/reviewer jiefu

iklam commented:
/reviewer credit jiefu

iklam commented:
/integrate
